commands/interceptor/duration.go

```diff
@@ -0,0 +1,140 @@
+	var possibleError error = nil
+	for step, layout := range stepFormats {
+		t, err := time.Parse(layout, unparsed)
+		if err == nil {
+			return step, t, nil
+		}
+		possibleError = err
+	}
+	return schema.StepSecond, time.Time{}, possibleError
+}
+
+// DurationInterceptor sets the duration if absent, and formats it accordingly,
+// see ParseDuration
+func DurationInterceptor(ctx *cli.Context) error {
+	start := ctx.String("start")
+	end := ctx.String("end")
+
+	startTime, endTime, step := ParseDuration(start, end)
+
+	if err := ctx.Set("start", startTime.Format(stepFormats[step])); err != nil {
+		return err
+	} else if err := ctx.Set("end", endTime.Format(stepFormats[step])); err != nil {
+		return err
+	} else if err := ctx.Set("step", step.String()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ParseDuration parses the `start` and `end` to a triplet, (startTime, endTime, step)
+// if --start and --end are both absent, then: start := now - 30min; end := now
+// if --start is given, --end is absent, then: end := now + 30 units, where unit is the precision of `start`, (hours, minutes, etc.)
+// if --start is absent, --end is given, then: start := end - 30 unis, where unit is the precision of `end`, (hours, minutes, etc.)
+// NOTE that when either(both) `start` or `end` is(are) given, there is no timezone info
+// in the format, (e.g. 2019-11-09 1001), so they'll be considered as UTC-based,
+// and generate the missing `start`(`end`) based on the same timezone, UTC
+func ParseDuration(start string, end string) (time.Time, time.Time, schema.Step) {
+	logger.Log.Debugln("Start time:", start, "end time:", end)
+
+	now := time.Now().UTC()
+
+	// both are absent
+	if len(start) == 0 && len(end) == 0 {
+		return now.Add(-30 * time.Minute), now, schema.StepMinute
+	}
+
+	var startTime, endTime time.Time
+	var step schema.Step
+	var err error
+
+	// both are present
+	if len(start) > 0 && len(end) > 0 {
+		start, end = AlignPrecision(start, end)
+
+		if _, startTime, err = tryParseTime(start); err != nil {
+			logger.Log.Fatalln("Unsupported time format:", start, err)
+		}
+		if step, endTime, err = tryParseTime(end); err != nil {
+			logger.Log.Fatalln("Unsupported time format:", end, err)
+		}
+
+		return startTime, endTime, step
+	}
+
+	// end is absent
+	if len(end) == 0 {
+		if step, startTime, err = tryParseTime(start); err != nil {
+			logger.Log.Fatalln("Unsupported time format:", start, err)
+		}
+		return startTime, startTime.Add(30 * stepDuration[step]), step
+	}
+
+	// start is present
+	if len(start) == 0 {
+		if step, endTime, err = tryParseTime(end); err != nil {
+			logger.Log.Fatalln("Unsupported time format:", end, err)
+		}
+		return endTime.Add(-30 * stepDuration[step]), endTime, step
+	}
+
+	logger.Log.Fatalln("Should never happen")
+
+	return startTime, endTime, step
+}
+
+// AlignPrecision aligns the two time strings to same precision
+// by truncating the more precise one
```

Comment:
I don't think truncating == `more precise`? You are truncating the minute level time to the hour level for example

Reply:
> I don't think truncating == `more precise`? You are truncating the minute level time to the hour level for example

truncating **the more precise one**, minute level time is more precise, so truncate it to hour level, less precise